### PR TITLE
Use transcript protein sequence

### PR DESCRIPTION
**Pull request by iskandr**
Overall motivation: move toward translating as few codons as necessary for each annotation (instead of always translating the full original & mutated proteins)

* Use `transcript.protein_sequence` instead of translating original protein
* Created special case for SNVs in `coding_effect` (plan to generalize this as a cleaner logic for all in-frame coding effects), translates only single affected codon
* Uses PyEnsembl 0.6.2

[Review on Reviewable](https://reviewable.io/reviews/hammerlab/varcode/45)

### CODE REVIEW

File: varcode/coding_effect.py [hunk 1]
```diff
@@ -32,12 +35,28 @@
...
 START_CODONS = set(CodonTable.standard_dna_table.start_codons)
 STOP_CODONS = set(CodonTable.standard_dna_table.stop_codons)
 
+def translate_codon(codon, aa_pos):
+    """Translate a single codon into a single amino acid or stop '*'
+
+    Parameters
+    ----------
+    codon : str
+        Expected to be of length 3
```
Comment (timodonnell) [inline varcode/coding_effect.py]: might as well `assert` this?

File: varcode/coding_effect.py [hunk 2]
```diff
@@ -32,12 +35,28 @@
...
+    """
+    # not handling rare Leucine or Valine starts!
+    if aa_pos == 0 and codon in START_CODONS:
+        return "M"
+    elif codon in STOP_CODONS:
+        return "*"
+    else:
+        return DNA_CODON_TABLE[codon]
+
 def translate(cds_seq):
```
Comment (timodonnell) [inline varcode/coding_effect.py]: super nitpicky nit: maybe rename `cds_seq` to something a little clearer like `coding_sequence` or even just `cds`.  If `cds` stand for "coding sequence", then `cds_seq` would be a sequence of coding sequences?

Also perhaps clarify the significance of it being a `cDNA coding sequence` in the comment? Does that mean that it's the reverse complement of the mRNA transcript, or is that meant to imply some more subtle restriction on what I can pass to this function?

Comment (timodonnell) [inline varcode/coding_effect.py]: Realizing that you use `cds_seq` in a number of places. If you prefer that name, that's fine with me.

File: varcode/util.py [hunk 3]
```diff
@@ -60,6 +64,15 @@ def random_variants(count, ensembl_release=MAX_ENSEMBL_RELEASE):
         if transcript.on_backward_strand:
             ref = reverse_complement(ref)
         alt_nucleotides = [x for x in VALID_NUCLEOTIDES if x != ref]
+        if insertions:
+            nucleotide_pairs = [
```
Comment (timodonnell) [inline varcode/util.py]: worth testing insertions > length 2?

File: varcode/util.py [hunk 4]
```diff
@@ -60,6 +64,15 @@ def random_variants(count, ensembl_release=MAX_ENSEMBL_RELEASE):
...
+                x + y
+                for x in alt_nucleotides
+                for y in VALID_NUCLEOTIDES
+            ]
+            alt_nucleotides.extend(nucleotide_pairs)
+        if deletions:
+            alt_nucleotides.append("")
```
Comment (timodonnell) [inline varcode/util.py]: likewise, maybe include longer deletions?

File: varcode/util.py [hunk 5]
```diff
@@ -28,7 +28,11 @@
 _transcript_ids_cache = {}
 
 
-def random_variants(count, ensembl_release=MAX_ENSEMBL_RELEASE):
+def random_variants(
```
Comment (timodonnell) [inline varcode/util.py]: consider setting a random seed (or taking one as an optional argument) so that the tests are deterministic? or does `random` module default to a fixed seed?

**Comment by timodonnell** [comment]
LGTM. I like the random variant annotation testing!

File: varcode/coding_effect.py (hunk 1 above)
Comment (iskandr) [inline varcode/coding_effect.py]: Asserting in a second branch that does refactoring on top of this.

File: varcode/coding_effect.py (hunk 2 above)
Comment (iskandr) [inline varcode/coding_effect.py]: What do you think of just using `cds`? It's pretty pervasive genomics and I agree that `cds_seq` is redundant.

File: varcode/util.py (hunk 3 above)
Comment (iskandr) [inline varcode/util.py]: Added an issue: https://github.com/hammerlab/varcode/issues/47

File: varcode/util.py (hunk 4 above)
Comment (iskandr) [inline varcode/util.py]: https://github.com/hammerlab/varcode/issues/47

File: varcode/util.py (hunk 5 above)
Comment (iskandr) [inline varcode/util.py]: https://github.com/hammerlab/varcode/issues/48

Also, since I'm currently using the non-determinism of this test to find edge cases, also going to add a commandline script for that purpose: 

https://github.com/hammerlab/varcode/issues/49

File: varcode/coding_effect.py (hunk 2 above)
Comment (timodonnell) [inline varcode/coding_effect.py]: Yeah, I have a mild preference for `coding_sequence` but I'd be fine with
`cds`.
